### PR TITLE
feat!: migrate from deprecated request module to axios

### LINKED ISSUE #1042: warning > request@2.88.2 deprecated
warning @sendgrid/mail > @sendgrid/client > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

### PR DESCRIPTION
**Pull request by eshanholtz**
Fixes #1042

### CODE REVIEW

**Comment by eshanholtz** [comment]
Upgrading `eslint` (security remediation) requires node >= 8

File: packages/client/src/classes/client.js [hunk 1]
```diff
@@ -112,21 +122,25 @@ class Client {
...
-        //Request error
-        if (error) {
+      axios(request)
+        .then(response => {
+          // Successful response
+          var parsedResponse = {
+            statusCode: response.status,
+            body: response.data,
+          };
+          return resolve([parsedResponse, response.data]);
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: This is a potentially breaking change so I'd like some feedback if we think there should be a major version roll.

Previously, this returned `[RequestResponse, any]` and now it's returning `[{body: any, statusCode: number}, any]` which does not match the TS which is `[AxiosResponse, any]`. Ideally, this should not be exposing the underlying type because it's an implementation detail that may change in the future. Similar to the request type.

I think the types should be defined in this repo and not expose the Axios/Request lib types.

File: package.json [hunk 2]
```diff
@@ -14,7 +14,7 @@
     "chai": "^4.0.1",
     "chai-as-promised": "^7.1.1",
     "dirty-chai": "^2.0.1",
-    "eslint": "^4.19.1",
+    "eslint": "^6.8.0",
```
Comment (childish-sambino) [inline package.json]: I don't like having to drop support for versions of Node in order to support a dev dependency. Prefer to get some usage numbers first to determine if it's okay to drop these first.

File: packages/client/src/classes/client.js (hunk 1 above)
Comment (eshanholtz) [reply packages/client/src/classes/client.js]: I agree. I'll implement an interface similar to [what we have in twilio-node](https://github.com/twilio/twilio-node/tree/master/lib/http). I also agree that this should warrant a major version roll

File: package.json (hunk 2 above)
Comment (childish-sambino) [reply package.json]: Can we use a 5.X version for now and/or move this to another PR?

File: packages/helpers/classes/index.d.ts [hunk 3]
```diff
@@ -2,12 +2,16 @@ import Attachment from "@sendgrid/helpers/classes/attachment";
 import EmailAddress from "@sendgrid/helpers/classes/email-address";
 import Mail from "@sendgrid/helpers/classes/mail"
 import Personalization from "@sendgrid/helpers/classes/personalization";
+import Request from "@sendgrid/helpers/classes/request";
 import ResponseError from "@sendgrid/helpers/classes/response-error";
+import Response from "@sendgrid/helpers/classes/response";
```
Comment (childish-sambino) [inline packages/helpers/classes/index.d.ts]: nit: `Response` before `ResponseError` here and elsewhere.

File: packages/client/src/client.d.ts [hunk 4]
```diff
@@ -26,12 +29,49 @@ declare class Client {
...
 
   /**
    * Do a request
    */
-  request(data: ClientRequest, cb?: (err: ResponseError, response: [ClientResponse, any]) => void): Promise<[ClientResponse, any]>;
+  request<TData>(opts: Client.RequestOptions<TData>, cb?: (err: ResponseError, response: [ClientResponse, any]) => void): Promise<[ClientResponse, any]>;
+}
+
+declare namespace Client {
+  export interface RequestOptions<TData = any, TParams = object> {
```
Comment (childish-sambino) [inline packages/client/src/client.d.ts]: This looks like mostly a duplicate of `Request.RequestOptions`.

File: packages/helpers/classes/request.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+var _ = require('lodash');
```
Comment (childish-sambino) [inline packages/helpers/classes/request.js]: Generally we're trying to move away from using lodash. Should eval if this can be dropped.

File: test.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+
+// using Twilio SendGrid's v3 Node.js Library
+// https://github.com/sendgrid/sendgrid-nodejs
+const sgMail = require('./packages/mail');
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+const msg = {
+  to: 'ea.shanholtz@gmail.com',
+  from: 'ea.shanholtz@gmail.com',
```
Comment (childish-sambino) [inline test.js]: Don't think this belongs.

File: package.json (hunk 2 above)
Comment (eshanholtz) [reply package.json]: Done: https://github.com/sendgrid/sendgrid-nodejs/pull/1063/files

File: packages/client/src/client.d.ts (hunk 4 above)
Comment (eshanholtz) [reply packages/client/src/client.d.ts]: So I've played around with this a bit, and haven't been able to import Request.RequestOptions. We have a similar duplication in `twilio-node`: [request.d.ts](https://github.com/twilio/twilio-node/blob/master/lib/http/request.d.ts) vs. [RequestClient.d.ts](https://github.com/twilio/twilio-node/blob/master/lib/base/RequestClient.d.ts)

Comment (eshanholtz) [reply packages/client/src/client.d.ts]: Figured out a workaround

File: package.json [hunk 7]
```diff
@@ -14,7 +14,7 @@
     "chai": "^4.0.1",
     "chai-as-promised": "^7.1.1",
     "dirty-chai": "^2.0.1",
-    "eslint": "^4.19.1",
+    "eslint": "^5.16.0",
```
Comment (childish-sambino) [inline package.json]: Should be part of a separate PR.

File: .travis.yml [hunk 8]
```diff
@@ -1,7 +1,6 @@
 language: node_js
 node_js:
   - 6
-  - 7
```
Comment (childish-sambino) [inline .travis.yml]: Should be part of a separate PR.

File: packages/client/src/classes/client.js [hunk 9]
```diff
@@ -72,7 +74,7 @@ class Client {
   createHeaders(data) {
 
     //Merge data with default headers
-    const headers = mergeData(this.defaultHeaders, data);
+    let headers = mergeData(this.defaultHeaders, data);
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: Why this change?

File: packages/client/src/classes/client.js [hunk 10]
```diff
@@ -86,47 +88,53 @@ class Client {
...
-
-    //Keep URL parameter consistent
-    if (data.uri) {
-      data.url = data.uri;
-      delete data.uri;
-    }
+  createRequest(opts) {
+
+    let options = {
+      url: opts.url || opts.uri,
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: For compatibility with above deleted block, should be swapped:
```
opts.uri || opts.url
```

File: packages/client/src/classes/client.js [hunk 11]
```diff
@@ -86,47 +88,53 @@ class Client {
...
+      baseURL: opts.baseURL || this.defaultRequest.baseURL,
+      method: opts.method || this.defaultRequest.method,
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: Isn't this what `mergeData` does?

File: packages/client/src/classes/client.js [hunk 12]
```diff
@@ -86,47 +88,53 @@ class Client {
...
     const promise = new Promise((resolve, reject) => {
-      http(request, (error, response, body) => {
-
-        //Request error
-        if (error) {
+      axios(opts)
+        .then(response => {
+          // Successful response
+          _this.lastResponse = new Response(response.status, response.data, response.headers);
+          return resolve([_this.lastResponse, response.data]);
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: This could result in threading issues if `lastResponse` get reassigned before returning.

File: packages/helpers/classes/attachment.js [hunk 13]
```diff
@@ -79,8 +79,7 @@ class Attachment {
     if (typeof content === 'string') {
       this.content = content;
       return;
-    }
-    else if (content instanceof Buffer && content.toString !== undefined) {
+    } else if (content instanceof Buffer && content.toString !== undefined) {
```
Comment (childish-sambino) [inline packages/helpers/classes/attachment.js]: Should be part of a separate PR.

File: packages/helpers/classes/attachment.js [hunk 14]
```diff
@@ -52,7 +52,7 @@ class Attachment {
 
     if ((typeof content !== 'undefined') && (typeof filePath !== 'undefined')) {
       throw new Error(
-        `The props 'content' and 'filePath' cannot be used together.`
+        'The props \'content\' and \'filePath\' cannot be used together.'
```
Comment (childish-sambino) [inline packages/helpers/classes/attachment.js]: Should be part of a separate PR.

File: packages/helpers/classes/index.d.ts [hunk 15]
```diff
@@ -2,12 +2,16 @@ import Attachment from "@sendgrid/helpers/classes/attachment";
...
   Attachment,
   EmailAddress,
   Mail,
   Personalization,
-  ResponseError,
-}
\ No newline at end of file
+  Request,
+  Response,
+  ResponseError
```
Comment (childish-sambino) [inline packages/helpers/classes/index.d.ts]: nit: Keep trailing `,`

File: packages/helpers/classes/request.d.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+type HttpMethod = 'get'| 'GET'|'post'|'POST'|'put'|'PUT'|'patch'|'PATCH'|'delete'|'DELETE';
```
Comment (childish-sambino) [inline packages/helpers/classes/request.d.ts]: nit: Spacing

File: packages/helpers/classes/request.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+class Request {
+  constructor(opts) {
+    opts = opts || {};
+    this.ANY = '*';
```
Comment (childish-sambino) [inline packages/helpers/classes/request.js]: What's the point/value of this?

File: package.json (hunk 7 above)
Comment (eshanholtz) [reply package.json]: It is: https://github.com/sendgrid/sendgrid-nodejs/pull/1063

File: packages/client/src/classes/client.js (hunk 9 above)
Comment (eshanholtz) [reply packages/client/src/classes/client.js]: Ah. Must have forgotten to put it back. Just playing around with the changes.

File: packages/helpers/classes/request.js (hunk 17 above)
Comment (eshanholtz) [reply packages/helpers/classes/request.js]: It was used for Request comparison, but since I took that out with lodash, I should take this out, too.

File: .travis.yml (hunk 8 above)
Comment (childish-sambino) [reply .travis.yml]: What I mean is the changes should be excluded from this PR.

File: packages/client/src/classes/client.js (hunk 12 above)
Comment (eshanholtz) [reply packages/client/src/classes/client.js]: I think I fixed this problem, can you double check?

File: packages/client/src/classes/client.js [hunk 18]
```diff
@@ -3,22 +3,22 @@
...
-const http = require('request');
+const axios = require('axios');
 const pkg = require('../../package.json');
 const {
   helpers: {
     mergeData,
   },
   classes: {
     ResponseError,
+    Response,
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: nit: ordering

File: packages/client/src/classes/client.js [hunk 19]
```diff
@@ -30,13 +30,14 @@ class Client {
...
+      'Content-Type': 'application/json',
       'User-agent': 'sendgrid/' + pkg.version + ';nodejs',
     };
 
     //Empty default request
     this.defaultRequest = {
-      json: true,
-      baseUrl: 'https://api.sendgrid.com/',
+      data: {},
+      baseURL: 'https://api.sendgrid.com/',
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: Prefer to make this `baseUrl` in the default request and then adapt it later for axios.

File: packages/client/src/classes/client.js [hunk 20]
```diff
@@ -62,7 +63,11 @@ class Client {
    * Set default request
    */
   setDefaultRequest(key, value) {
-    this.defaultRequest[key] = value;
+    if (key === 'baseUrl') {
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: Then this logic isn't needed.

File: packages/client/src/classes/client.js [hunk 21]
```diff
@@ -86,47 +91,52 @@ class Client {
...
-    }
+  createRequest(opts) {
+
+    let options = {
+      url: opts.uri || opts.url,
+      baseURL: opts.baseUrl,
+      method: opts.method,
+      data: opts.body || {},
+      params: opts.qs || {},
+      headers: opts.headers || {},
```
Comment (childish-sambino) [inline packages/client/src/classes/client.js]: Is this `||`ing here needed?

File: packages/client/src/request.d.ts [hunk 22]
```diff
@@ -1,3 +1,4 @@
-import {OptionsWithUrl} from "request";
+import { Request, RequestOptions } from "@sendgrid/helpers/classes/request";
```
Comment (childish-sambino) [inline packages/client/src/request.d.ts]: nit: Spacing around curly brackets is not consistent between files.

File: packages/helpers/classes/response-error.js [hunk 23]
```diff
@@ -19,12 +19,12 @@ class ResponseError extends Error {
...
-    const {headers, statusCode, statusMessage, body} = response;
+    const {headers, status, statusText, data} = response;
 
     //Set data
-    this.code = statusCode;
-    this.message = statusMessage;
-    this.response = {headers, body};
+    this.code = status;
+    this.message = statusText;
+    this.response = {headers, data};
```
Comment (childish-sambino) [inline packages/helpers/classes/response-error.js]: Should be `{headers, body: data}` for compatibility.

File: packages/helpers/classes/request.js [hunk 24]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+class Request {
```
Comment (childish-sambino) [inline packages/helpers/classes/request.js]: Is this class still needed?

File: packages/mail/src/classes/mail-service.js [hunk 25]
```diff
@@ -192,7 +192,7 @@ class MailService {
       const request = {
         method: 'POST',
         url: '/v3/mail/send',
-        body,
+        data: body,
```
Comment (childish-sambino) [inline packages/mail/src/classes/mail-service.js]: Don't think this should change.

File: test/typescript/client.ts [hunk 26]
```diff
@@ -26,5 +26,5 @@ const req = Client.createRequest({
 Client.request({
   url: "/test"
 }).then(res => {
-  res[0].headers
-})
\ No newline at end of file
+  res[0].statusCode;
```
Comment (childish-sambino) [inline test/typescript/client.ts]: From `headers` to `statusCode`?

File: packages/helpers/classes/request.js (hunk 24 above)
Comment (eshanholtz) [reply packages/helpers/classes/request.js]: I guess we only need the typescript one

File: test/typescript/client.ts (hunk 26 above)
Comment (eshanholtz) [reply test/typescript/client.ts]: Shouldn't matter, I was making sure the typescript response fields were populating properly. and that I had the right variable names.

File: packages/client/src/classes/client.js (hunk 19 above)
Comment (eshanholtz) [reply packages/client/src/classes/client.js]: Where is an appropriate "later"? Also, if this dict is meant to be merged with the renamed opts in the `request` function, then the `defaultRequest` should have keys that align with what's adapted for axios

File: packages/client/src/classes/client.js (hunk 21 above)
Comment (eshanholtz) [reply packages/client/src/classes/client.js]: Nope

File: packages/client/src/classes/client.js (hunk 19 above)
Comment (eshanholtz) [reply packages/client/src/classes/client.js]: @childish-sambino can I get some feedback?

Comment (childish-sambino) [reply packages/client/src/classes/client.js]: I think it should be adapted after L105 where the data is merged. We define the structure of default request, merge it with the options provided in the request, and then adapt it to be whatever the underlying client is expecting.

File: packages/client/package.json [hunk 27]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@sendgrid/client",
   "description": "Twilio SendGrid NodeJS API client",
-  "version": "6.5.5",
+  "version": "6.5.3",
```
Comment (childish-sambino) [inline packages/client/package.json]: Needs to be sync'd.

File: packages/helpers/classes/response-error.js [hunk 28]
```diff
@@ -40,10 +40,10 @@ class ResponseError extends Error {
    * Convert to string
    */
   toString() {
-    const {body} = this.response;
+    const {data} = this.response;
```
Comment (childish-sambino) [inline packages/helpers/classes/response-error.js]: This should be reverted back to `body`.

File: packages/helpers/classes/request.d.ts [hunk 29]
```diff
@@ -0,0 +1,15 @@
...
+    method?: HttpMethod | '*';
+    url?: string;
+    auth?: string;
+    params?: TParams;
+    data?: TData | '*';
+    headers?: object | '*';
+    ca?: string;
+}
+
+export class Request<TData = any> {
```
Comment (childish-sambino) [inline packages/helpers/classes/request.d.ts]: This class doesn't exist anymore in the JS.

File: packages/helpers/classes/request.d.ts [hunk 30]
```diff
@@ -0,0 +1,15 @@
+type HttpMethod = 'get'|'GET'|'post'|'POST'|'put'|'PUT'|'patch'|'PATCH'|'delete'|'DELETE';
+
+export interface RequestOptions<TData = any, TParams = object> {
+    method?: HttpMethod | '*';
+    url?: string;
+    auth?: string;
+    params?: TParams;
+    data?: TData | '*';
+    headers?: object | '*';
+    ca?: string;
```
Comment (childish-sambino) [inline packages/helpers/classes/request.d.ts]: Should only contain `url`, `baseUrl`, `method`, `body`, `qs`, and `headers`.

File: packages/helpers/classes/response-error.js (hunk 28 above)
Comment (eshanholtz) [reply packages/helpers/classes/response-error.js]: This uses the raw response object from `axios`

Comment (childish-sambino) [reply packages/helpers/classes/response-error.js]: No, it's using the response object defined in L27.

File: packages/client/src/request.d.ts [hunk 31]
```diff
@@ -1,3 +1,3 @@
-import {OptionsWithUrl} from "request";
+import RequestOptions from "@sendgrid/helpers/classes/request";
 
-export type ClientRequest = OptionsWithUrl;
\ No newline at end of file
+export default RequestOptions;
```
Comment (childish-sambino) [inline packages/client/src/request.d.ts]: Can we just keep this a `ClientRequest` for compatibility (and less changes)?

Comment (eshanholtz) [reply packages/client/src/request.d.ts]: Done, also put back some other unnecessary changes

**Comment by jandppw** [comment]
Can somebody specify _why_ this is breaking? (So we can check whether it has impact on our usage code). I can't see that from browsing the code changes.

**Comment by childish-sambino** [comment]
@jandppw https://github.com/sendgrid/sendgrid-nodejs/blob/master/docs/migration-guides/migrating-from-version-6-to-7.md
